testing/integration/controller-wifi.nix

```diff
@@ -0,0 +1,331 @@
+                guest.port = 3333;
+            }
+        ];
+
+        # === dnsmasq + static IPs for wlan0*
+
+        # Setup networking + DHCP (dnsmasq) for the simulated wireless APs.
+        # Without this connman will wait ages for a (non-existant) DHCP
+        # offer, makin the tests unbearably slow.
+        #
+        # Note: this could probably be made to properly route traffic via the
+        # test VLANs or QEMU's vnet, but for now just returns some values in
+        # the 10.0.9.0/24 subnet
+        networking.interfaces =
+            let staticIpCfg = {
+                ipv4.addresses = [{
+                    address = "10.0.9.1"; # irrelevant
+                    prefixLength = 24;
+                }];
+            };
+            in
+            attrsets.genAttrs allSimulatedAPInterfaces (_: staticIpCfg);
+
+        services.dnsmasq.enable = true;
+        services.dnsmasq.settings = {
+            interface = allSimulatedAPInterfaces;
+
+            dhcp-option = [
+                "3,10.0.9.2" # gateway, does not exist
+                "6,10.0.9.3" # DNS, does not exist
+            ];
+            dhcp-range = "10.0.9.30,10.0.9.99,1h";
+        };
+
+        # disable the local DNS caching proxy in connman to avoid conflicts with
+        # dnsmasq
+        services.connman.extraFlags = [ "--nodnsproxy" ];
+
+
+        # === Simulated wireless access points
+
+        # enable 802.11 simulation
+        boot.kernelModules = [ "mac80211_hwsim" ];
+
+        # wireless access points
+        services.hostapd = {
+          enable = true;
+          # note: do not change this to wlan1 or other id, weird failures appear
+          radios.wlan0 = {
+            band = "2g";
+            countryCode = "US";
+            networks = {
+              wlan0 = {
+                ssid = "test-ap-sae";
+                bssid = "02:00:00:00:00:00";
+                authentication = {
+                  mode = "wpa3-sae";
+                  saePasswords = [ { password = "reproducibility"; } ];
+                };
+              };
+              wlan0-1 = {
+                ssid = "test-ap-mixed";
+                bssid = "02:00:00:00:00:01";
+                authentication = {
+                  mode = "wpa3-sae-transition";
+                  saeAddToMacAllow = true;
+                  saePasswordsFile = pkgs.writeText "password" "reproducibility";
+                  wpaPasswordFile = pkgs.writeText "password" "reproducibility";
+                };
+              };
+
+              # connman hangs when connecting to this AP, nothing in the logs
+              # neither from connman nor from wpa_supplicant ¯\_(ツ)_/¯
+              #wlan0-2 = {
+              #  ssid = "test-ap-wpa2";
+              #  authentication = {
+              #    mode = "wpa2-sha256";
+              #    wpaPassword = "reproducibility";
+              #  };
+              #  bssid = "02:00:00:00:00:02";
+              #};
```

Comment:
I would be curious whether this issue persists with connman 1.43, because it sounds exactly like the behavior I saw with the paid network in my area (though that was unauthenticated + captive portal, so not an obvious match).

Reply:
I just re-ran the tests with a nixpkgs bump merged in, and it completes when this is commented out, but fails when this AP is included.

Pity!

Reply:
As long as we can not resolve this, shall we just more concisely document that mode `wpa2-sha256` seems to have issues?

A more drastic (and possibly correct) measure would be to deny connection requests to such networks and make them officially unsupported. They seem to be rare, with generally low support, and WPA3 is an alternative nowadays. But I guess we can not (easily) detect this from connman?

Reply:
OK, how about adding some context then that helps re-evaluate in the future?

Maybe just prefix the explanation with `As of connman 1.43, it hangs ...`?

Reply:
Done in 4d62603